src/rgw/rgw_kms.cc

```diff
@@ -17,12 +18,283 @@
+
+  virtual ~VaultSecretEngine(){}
+};
+
+
+class TransitSecretEngine: public VaultSecretEngine {
+
+private:
+  int get_key_version(boost::string_view key_id, string& version)
+  {
+    vector<boost::string_view> tokens;
+    size_t pos = 0;
+
+    pos = key_id.rfind("/", key_id.length());
+    if (pos != std::string::npos){
+      boost::string_view token = key_id.substr(pos, key_id.length());
+      if (tokens.size() >= 1 && token.find_first_not_of("0123456789") == std::string::npos){
+        version.assign(std::string(token));
+        return 0;
+      }
+    }
+    return -1;
+  }
+
+public:
+  TransitSecretEngine(CephContext *cct): VaultSecretEngine(cct){ }
+
+  int get_key(boost::string_view key_id, std::string& actual_key)
+  {
+    JSONParser* parser = new JSONParser();
+    string version;
+
+    if (get_key_version(key_id, version) < 0){
+      ldout(cct, 20) << "Missing or invalid key version" << dendl;
+      return -EINVAL;
+    }
+
+    int res = send_request(key_id, parser);
+    if (res < 0) {
+      return res;
+    }
+
+    JSONObj* json_obj = &(*parser);
+    std::array<std::string, 3> elements = {"data", "keys", version};
+    for(const auto& elem : elements) {
+      json_obj = json_obj->find_obj(elem);
+      if (!json_obj) {
+        ldout(cct, 0) << "ERROR: Key not found in JSON response from Vault using Transit Engine" << dendl;
+        return -EINVAL;
+      }
+    }
+
+    return decode_secret(json_obj, actual_key);
+  }
+
+};
+
+class KvSecretEngine: public VaultSecretEngine {
+
+public:
+
+  KvSecretEngine(CephContext *cct): VaultSecretEngine(cct){ }
+
+  virtual ~KvSecretEngine(){}
+
+  int get_key(boost::string_view key_id, std::string& actual_key){
+    JSONParser parser;
+    int res = send_request(key_id, &parser);
+    if (res < 0) {
+      return res;
+    }
+
+    JSONObj *json_obj = &parser;
+    std::array<std::string, 3> elements = {"data", "data", "key"};
+    for(const auto& elem : elements) {
+      json_obj = json_obj->find_obj(elem);
+      if (!json_obj) {
+        ldout(cct, 0) << "ERROR: Key not found in JSON response from Vault using KV Engine" << dendl;
+        return -EINVAL;
+      }
+    }
+    return decode_secret(json_obj, actual_key);
+  }
+
+};
+
+
+class VaultKMS : KMS {
+
+protected:
+  CephContext *cct;
+  VaultSecretEngine *engine;
+
+public:
+  int get_key(boost::string_view key_id, std::string& actual_key) {
+    if (!engine){
+      ldout(cct, 5) << "ERROR: No supported Vault Secret Engine found" << dendl;
+      return -EINVAL;
+    }
+    return engine->get_key(key_id, actual_key);
+  }
+
+  VaultKMS(CephContext *cct) {
+    std::string secret_engine = cct->_conf->rgw_crypt_vault_secret_engine;
+
+    ldout(cct, 10) << "Selected " << secret_engine << " secret engine" << dendl;
+
+    if (RGW_SSE_KMS_VAULT_SE_KV == secret_engine)
+      engine = new KvSecretEngine(cct);
+    else if (RGW_SSE_KMS_VAULT_SE_TRANSIT == secret_engine)
+      engine = new TransitSecretEngine(cct);
```

Comment:
this design imposes some unnecessary heap allocations on every request, and we aren't really taking advantage of the polymorphism that the KMS and SecretEngine interfaces provide

i would prefer a model where everything lives on the stack in `get_actual_key_from_vault()`, ie:
```c++
  std::string secret_engine = cct->_conf->rgw_crypt_vault_secret_engine;

  ldout(cct, 10) << "Selected " << secret_engine << " secret engine" << dendl;

  if (RGW_SSE_KMS_VAULT_SE_KV == secret_engine) {
    KvSecretEngine engine(cct);
    return engine.get_key(key_id, actual_key);
  } else if (RGW_SSE_KMS_VAULT_SE_TRANSIT == secret_engine) {
    TransitSecretEngine engine(cct);
    return engine.get_key(key_id, actual_key);
  } else {
    ldout(cct, 5) << "ERROR: Unsupported rgw_crypt_vault_secret_engine=" << secret_engine << dendl;
    return -EINVAL;
  }
```
this simplifies out the KMS interface entirely, and still allows the SecretEngine interface to be mocked for unit testing. what do you think?